garage/misc/git_handler.py

```diff
@@ -0,0 +1,330 @@
+        return curr_branch_name
+
+    def run_experiment_on_target(self, target, exp_args):
+        """Run a experiment on the indicated commit defined by target.
+
+        After the target is verified, if there are any non-staged changes in
+        the working directory, they're stashed, the target commit is checked
+        out, the experiment is executed and the working directory is restored
+        at the end, including the stashed changes.
+
+        Parameters
+        ----------
+        - target: a key-value string, where the key indicates the type of
+            target (tag, branch, sha), and the value the reference for the
+            corresponding type. For example:
+                sha: 8d54s23a
+                branch: my_branch
+                branch: remote/master
+                tag: my_tag
+        - exp_args: it's a dictionary of arguments obtained from the function
+            run_experiment at garage.misc.instrument, that will be passed to
+            the experiment once the target is checked out.
+
+        Precondition
+        ------------
+        - The working directory is not in the middle of a conflict resolution.
+
+        """
+        assert isinstance(target, str), ("The git target has to be a string")
+        self.target = target
+        self.target_sha = self._verify_target(self.target)
+        self.restore_sha = self.get_sha()
+        self.stash_sha = self.create_stash()
+        self.branch_name = self.get_current_branch_name()
+        self.target_branch_name = ("run_exp_on_" + str(int(time.time())))
+        try:
+            self._switch_to_target(self.target_branch_name, self.target_sha)
+            current_sha = self.get_sha()
+            self._run_experiment(exp_args)
+        except BaseException:
+            print(colored(
+                "If the working area is damaged after " \
+                "an unexpected error while trying to\nswitch to the " \
+                "target in the repository, run the following command(s) " \
+                "to\nrestore it:", "yellow"))
+            if self.branch_name:
+                print(
+                    colored("$ git checkout %s" % (self.branch_name),
+                            "yellow"))
+            else:
+                print(
+                    colored("$ git reset --hard %s" % (self.restore_sha),
+                            "yellow"))
+
+            if self.stash_sha:
+                print(
+                    colored("$ git stash apply %s" % (self.stash_sha),
+                            "yellow"))
+            raise
+        finally:
+            self._restore_working_dir(self.target_branch_name, self.stash_sha,
+                                      self.branch_name, self.restore_sha)
+
+    def _verify_target(self, target):
+        """Verify the key-value target is valid.
+
+        Parameters
+        ----------
+        - target: same description as in method run_experiment_on_target.
```

Comment:
I think it's better to just copy-paste the actual description, rather than have people scroll back up to see what it was in the first place.

Reply:
Done.